Propagate sources in VariableBuilder and add SuperSource

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90575
* __->__ #91729

**Motivation**
When adding support for default args (#90575), a lot of VariableTrackers missing sources were encountered.  Currently, in a lot of cases it seems OK to skip the source for VariableTrackers created (especially during inlining), but that assumption breaks down when inlining functions with default arguments. 

**Summary** of changes
- propagate the self.source of the VariableBuilder to the new variables being built, which seems like it was an omission previously
- Add SuperSource to track usages of super(), so that SuperVariables can support function calls with default args


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Propagate sources in places where they were previously missing

Sometimes, we don't have a valid source to propagate, but
when we do it's better to propagate it since downstream
something (such as default method args) could require it.

[ghstack-poisoned]

wconstab committed:
Update on "Propagate sources in places where they were previously missing"


Sometimes, we don't have a valid source to propagate, but
when we do it's better to propagate it since downstream
something (such as default method args) could require it.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Propagate sources in places where they were previously missing"


Sometimes, we don't have a valid source to propagate, but
when we do it's better to propagate it since downstream
something (such as default method args) could require it.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented on torch/_dynamo/side_effects.py:
```diff
@@ -211,7 +211,14 @@ def track_object_new(
-            obj, mutable_local=AttributeMutationNew(None, cls_source), **options
+            obj,
+            mutable_local=AttributeMutationNew(None, cls_source),
+            # TODO(whc) this is a hack, i'm not sure what to do.
```
@voznesenskym @ezyang any idea here?

wconstab replied:
ok i figured this out (partly)

1) cls_source is definitely wrong here, since that represents the class not the instance, and that's just wrong.  passing cls_source is the cause of the `event_dim` problem TODO'd in guards.py in this PR.  So at least both mysteries are solved.

2) what can we do instead?  (A) is to not pass a source here, and then make carveouts in a bunch of places where we allow source to be None and propagate None in those cases.  TBD whether this really works or runs into trouble.  (B) is to come up with the notion of an 'InlinedVariableSource' that somehow ties in with the side effect machinery so that if we need to guard on something we can codegen the temporary variable.

@voznesenskym any idea about whether (B) is a good option?

ezyang commented:
this PR looks long and annoying to review, I probably won't get to it until I'm off vacation lol

wconstab committed:
Update on "Propagate sources in places where they were previously missing"


Sometimes, we don't have a valid source to propagate, but
when we do it's better to propagate it since downstream
something (such as default method args) could require it.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Propagate sources in VariableBuilder and add SuperSource"


**Motivation**
When adding support for default args (#90575), a lot of VariableTrackers missing sources were encountered.  Currently, in a lot of cases it seems OK to skip the source for VariableTrackers created (especially during inlining), but that assumption breaks down when inlining functions with default arguments. 

**Summary** of changes
- propagate the self.source of the VariableBuilder to the new variables being built, which seems like it was an omission previously
- Add SuperSource to track usages of super(), so that SuperVariables can support function calls with default args


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Propagate sources in VariableBuilder and add SuperSource"


**Motivation**
When adding support for default args (#90575), a lot of VariableTrackers missing sources were encountered.  Currently, in a lot of cases it seems OK to skip the source for VariableTrackers created (especially during inlining), but that assumption breaks down when inlining functions with default arguments. 

**Summary** of changes
- propagate the self.source of the VariableBuilder to the new variables being built, which seems like it was an omission previously
- Add SuperSource to track usages of super(), so that SuperVariables can support function calls with default args


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/source.py:
```diff
@@ -251,6 +251,30 @@ def name(self):
+        return self.obj.guard_source()
+
+    def name(self):
+        return f"super({self.type.name()}, {self.obj.name()})"
```
Why can't this just be super(); can super be materially different in Python 3?

wconstab replied:
i think technically you can pass another class as the second obj in the super call, and you'll run MRO but skip some steps.  I don't think this is a common use case, but i figured we might as well capture the super call precisely as it was issued.

ezyang reviewed (approved):
Alright, I finally grokked the pattern to the changes and it looks great. Don't feel blocked by the super comment; the way its written may very well be this way. 

I do wonder how we can get people to properly propagate source as needed. Maybe we should make it a mandatory kwarg and have an easily greppable string for passing None

wconstab commented:
> I do wonder how we can get people to properly propagate source as needed. Maybe we should make it a mandatory kwarg and have an easily greppable string for passing None

@ezyang yea i hadn't thought of doing it this way, but i could try it, (if you explain it better)

another way is to actually assert source in more places.  i was hoping to go further in this direction, but it is not very realistic unless we can create a source for everything that's constructed during Inlining.  

It might be possible to create an 'Inlined' source, which is valid to be passed to derived sources like AttrSource, but which doesn't do anything until used in a guard. If used in a guard, it could force the translator to assign a local name to the temporary, or something.  I haven't thought this through.  Wonder what others think here?

wconstab commented:
@pytorchbot merge

ezyang commented:
> yea i hadn't thought of doing it this way, but i could try it, (if you explain it better)

Well, I'm not exactly sure how it would work. But it is similar to the assert strategy, just that if you have `def f(*, source)` (no default arg for source), we will raise a runtime error if you call f without passing source=

> it is not very realistic unless we can create a source for everything that's constructed during Inlining.

Hmmmf. Now I'm wondering if we even did the morally right thing in this PR.

In general, we don't expect to have a source for intermediates inside of a Dynamo region. Because the point about Source is to talk about quantities that exist in the ambient Python runtime; we don't want to do compute inside a Source because then we'd keep recomputing inside our guard evaluation. So there is something deeply suspicious about inlining a function requiring Sources to be defined, because all of the function arguments could have been intermediates, so really nothing should require Sources to be defined at this point. This goes back to

> it seems OK to skip the source for VariableTrackers created (especially during inlining), but that assumption breaks down when inlining functions with default arguments.

I'd kind of like to understand how this assumption breaks down in more detail.

cc @jansel

jansel commented:
Note I haven't looked at this PR in detail, so just responding to the discussion.

Anything that can be computed without running any PyTorch operations should have a source. So all graph inputs should have sources, while intermediates and graph outputs should not have sources.

For default args (that are not constants) the source could be something like:
```
fn_name.__defaults__[4]
```

wconstab commented:
@jansel I had to do a little reproing to figure one out, but I wrote something up on a separate PR.
https://github.com/pytorch/pytorch/pull/92091

I think that defaults is special, since it is an instance that lives on the class object which is not something we created during inlining.

jansel commented:
I don't see defaults in that example.  In that case we should be able to get the `__class__` directly without going through `AttrSource(self.source, "__class__")`.
